datafusion/physical-expr/src/expressions/binary.rs

```diff
@@ -48,11 +47,11 @@ use kernels::{
 };
 
 /// Binary expression
-#[derive(Debug, Hash, Clone)]
-pub struct BinaryExpr {
-    left: Arc<dyn PhysicalExpr>,
+#[derive(Debug, Hash, Clone, Eq, PartialEq)]
+pub struct BinaryExpr<DynPhysicalExpr: ?Sized = dyn PhysicalExpr> {
```

Comment:
The removal of the boiler plate in this PR is great ❤️ 

I feel like this additional trait / workaround is non ideal as it makes understanding the `BinaryExpr` very unobvious (there is now a level of indirection that is irrelevant for most of users of `BinaryExpr`).

Reply:
I am playing around with it to see if I can avoid it somehow

Reply:
Here is one way that seems to work: https://github.com/peter-toth/datafusion/pull/5

More verbose, but I think it keeps the structs simpler to understand

Reply:
Sure, we don't need to keep the generic parameter workaround. I agree that those params look a bit weird.
https://github.com/peter-toth/datafusion/pull/5 looks good to me, if you extend it to other expressions I can merge it tomorrow and amend my PR description.

Reply:
I think I accidentally push the commits to your branch 😬 -- hopefully that is ok

Reply:
No problem. :grin: I've just updated the PR description.